service/storage_proxy.cc

```diff
@@ -4593,6 +4595,39 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
     co_return condition_met;
 }
 
+bool storage_proxy::is_likely_to_respond_in_time(const gms::inet_address& ep, clock_type::time_point timeout) {
+    clock_type::time_point now = clock_type::now();
+    if (timeout <= now) {
+        return false;
+    }
+    if (fbu::is_me(ep)) {
+        return true;
+    }
+    if (auto last_seen_ep = last_seen(ep)) {
+        auto time_left = timeout - now;
+        auto last_attempt = now - last_seen_ep->last_sent;
+        // If the node has never responded yet or there was no communication attempt for a long while,
+        // the node should be probed in case it becomes responsive
+        if (last_seen_ep->last_responded == last_seen_info::unknown || last_attempt > time_left) {
+            return true;
+        }
+        auto time_without_response = last_seen_ep->last_sent - last_seen_ep->last_responded;
```

Comment:
After reading this code after a month, I'm confused... doesn't sounds like a right calculation indeed. It gets bogus results each time stop sending for a couple of seconds, and then send again but then check before the replica has a chance to respond. I'll try to figure out what I had in mind and either provide a meaningful explanation or rewrite the whole thing. "Time without response" is literally what you described in the comment, not what is calculated here.

Reply:
Oh wait, now, after I responded to your other comment, I remember more. I'm not really interested in `time_without_response` as in "how long ago we received a response from a node", but rather "how long did it take last time for this node to respond" - the name is confusing and unfortunate indeed. The problem is that I also want to estimate this value with as few bytes used and cpu cycles wasted, so it's only a very rough estimate. This formula is supposed to work well if the system is under constant load - and so we also very frequently try to contact all the replicas. If we don't contact the replicas frequently, the other mechanism (which you also asked about in a previous comment) checks if the information is not too stale - and if it is, it's dropped. With these two checks in place, it's possible to roughly guess if the node is responsive or not, but at the same time do not keep track of latencies of each request - because that would have been much costlier to maintain than a single pair of timestamps per endpoint. So it's entirely possible (and expected) that `last_sent` and `last_responded` do not actually come from the same request, but if the coordinator node is sufficiently busy, it's still enough to examine the values to detect if a target node might be dead - which would be evident after seeing lots of sending attempts, but no responses.

Reply:
Still, I'm not fully utilizing the information I have. Instead of just collecting `last_sent` and `last_responded` from potentially unrelated requests, it's easy enough to register the actual time it took for a full round trip when registering a response time - since when the response is registered, we still know when it was sent. Using that would simplify these questionable calculations and hopefully make the review clearer as well. I'll prepare a better v2.